pkg/controllers/termination/terminator/terminator.go

```diff
@@ -59,60 +61,61 @@ func (t *Terminator) Cordon(ctx context.Context, node *v1.Node) error {
 // https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown
 func (t *Terminator) Drain(ctx context.Context, node *v1.Node) error {
-	// Get evictable pods
-	pods, err := t.getPods(ctx, node)
-	if err != nil {
-		return fmt.Errorf("listing pods for node, %w", err)
+	// Grace period is up, we can no longer drain
+	if t.clock.Now().After(node.DeletionTimestamp.Time) {
+		return nil
 	}
-	var podsToEvict []*v1.Pod
-	// Skip node due to pods that are not able to be evicted
-	for _, p := range pods {
-		// Ignore if unschedulable is tolerated, since they will reschedule
-		if podutil.ToleratesUnschedulableTaint(p) {
-			continue
-		}
-		// Ignore static mirror pods
-		if podutil.IsOwnedByNode(p) {
-			continue
-		}
-		podsToEvict = append(podsToEvict, p)
-	}
-	// Enqueue for eviction
-	t.evict(podsToEvict)
-
-	if len(podsToEvict) > 0 {
-		return NewNodeDrainError(fmt.Errorf("%d pods are waiting to be evicted", len(podsToEvict)))
-	}
-	return nil
-}
 
-// getPods returns a list of evictable pods for the node
-func (t *Terminator) getPods(ctx context.Context, node *v1.Node) ([]*v1.Pod, error) {
 	podList := &v1.PodList{}
 	if err := t.kubeClient.List(ctx, podList, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
-		return nil, fmt.Errorf("listing pods on node, %w", err)
+		return fmt.Errorf("listing pods on node, %w", err)
 	}
-	var pods []*v1.Pod
-	for _, p := range podList.Items {
-		// Ignore if the pod is complete and doesn't need to be evicted
-		if podutil.IsTerminal(lo.ToPtr(p)) {
-			continue
+	pods := lo.Map(podList.Items, func(pod v1.Pod, _ int) *v1.Pod { return lo.ToPtr(pod) })
+	evictable := lo.Reject(pods, func(p *v1.Pod, _ int) bool {
+		return podutil.IsTerminal(p) ||
+			(podutil.IsTerminating(p) && t.clock.Now().After(p.DeletionTimestamp.Time.Add(1*time.Minute))) ||
+			podutil.ToleratesUnschedulableTaint(p) ||
+			podutil.IsOwnedByNode(p)
+	})
+	// We no longer have time to evict while still respecting pod
+	// TerminationGracePeriodSeconds, so taint the node with NoExecute.
+	if t.withinGracefulShutdownPeriod(node, pods) || len(evictable) == 0 {
+		if err := t.taintNoExecute(ctx, node); err != nil {
+			return fmt.Errorf("tainting no execute, %w", err)
 		}
-		// Ignore if kubelet is partitioned and pods are beyond graceful termination window
-		if t.isStuckTerminating(lo.ToPtr(p)) {
-			continue
+		return NewNodeDrainError(fmt.Errorf("%d pods are gracefully shutting down", len(pods)))
+	}
+	// Evict remaining pods
+	t.evict(evictable)
+	return NewNodeDrainError(fmt.Errorf("%d pods are waiting to be evicted", len(evictable)))
+}
+func (t *Terminator) withinGracefulShutdownPeriod(node *v1.Node, pods []*v1.Pod) bool {
+	gracePeriodSeconds := lo.Max(lo.Map(pods, func(pod *v1.Pod, _ int) int64 { return lo.FromPtr(pod.Spec.TerminationGracePeriodSeconds) }))
+	gracePeriod := time.Duration(gracePeriodSeconds) * time.Second
+	return t.clock.Now().Add(gracePeriod).After(node.DeletionTimestamp.Time)
```

Comment:
Just posting here to continue this in the review:

I think we still need to align on the philosophy here. The two philosophies are:
1. **[Your solution]** A Machine's DeletionTimestamp dictates the time it should no longer be present in the APIServer, so it should also include the GracefulTerminationPeriodSeconds of the pods that are scheduled to it. This means that we need a pod's deletion timestamp to line up with the node's deletion timestamp, where pods that haven't begun scaling down due to Eviction Errors with PDBs will receive a SIGINT at time: `MachineDeletionTimestamp.Time() - Pod.TerminationGracePeriodSeconds`.  
- If a pod's GracePeriod is >= machine GracePeriod, that means that we'll immediately add the NoExecute taint, beginning force deletion of all pods on the node (since TaintEviction doesn't respect PDBs). The further the maxPodGracePeriodSeconds is from the machine Grace Period, the longer it'll take for those pods to be forcibly removed.

2. A Machine's GracePeriodSeconds (which is used to calculate the DeletionTimestamp) dictates how long the object can be gracefully deleted before forceful actions are taken. In this case, we wouldn't try to sync up the pod's and machine's deletion timestamps, as they're not linked by any OwnerReferences/CascadeDeletion. Karpenter could take the Machine's GracePeriodSeconds to gracefully evict as many pods as possible, then initiate force deletion of all pods, respecting each of their GracefulTerminationPeriodSeconds, rather than initiating force deletion earlier for pods due to the GracefulTerminationPeriodSeconds of other pods, more likely violating PDBs.

I'm a fan of option 2, but curious what yours and other's thoughts are.

Reply:
Solution 1 violates PDBs. Solution 2 violates GraceTerminationPeriodSeconds. In my mind, it's more important to give the process time to clean up, than it is to wait for a replacement to come online. Obviously, we want to respect both, but if we have to choose, I think GraceTerminationPeriodSeconds is more essential. Think about the batch use case, where the ML training job needs to checkpoint and save its work, but there may not be a PDB.

Reply:
In the case of an ML training job without a PDB, it should start being cleaned up as soon as the DeletionTimestamp gets set, as Karpenter will evict all pods that don't have PDBs, predicting here won't start the cleanup any sooner. 

Like you said, being predictive helps out on enforcing the Machine's GraceTerminationPeriodSeconds, yet I think users really care more about enforcing the proper cleanup of their pods, and enforcement of availability requirements (not violating PDBs). I'd prefer we violate the Machine's GraceTerminationPeriodSeconds in this case since the implication here is paying more for an instance, but respecting the user's requirements on availability/graceful termination for their pods, which is what they're paying for.